### ISSUE TITLE
IMAP4Client breaks if server uses CONDSTORE

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2695]
|<img alt="slamb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| slamb reported|
|-|-|
|Trac ID|trac#2695|
|Type|defect|
|Created|2007-06-07 02:57:42Z|

My Cyrus server's capabilities line looks like this:

```
* OK [CAPABILITY IMAP4 IMAP4rev1 LITERAL+ ID STARTTLS] hobbes.slamb.org Cyrus IMAP4 v2.3.7-Invoca-RPM-2.3.7-4.centos4.slamb server ready
1 capability
* CAPABILITY IMAP4 IMAP4rev1 LITERAL+ ID STARTTLS ACL RIGHTS=kxte QUOTA MAILBOX-REFERRALS NAMESPACE UIDPLUS NO_ATOMIC_RENAME UNSELECT CHILDREN MULTIAPPEND BINARY SORT SORT=MODSEQ THREAD=ORDEREDSUBJECT THREAD=REFERENCES ANNOTATEMORE CATENATE CONDSTORE IDLE LISTEXT LIST-SUBSCRIBED X-NETSCAPE URLAUTH
1 OK Completed
```

Note the RFC:4551 stuff: `CONDSTORE SORT SORT=MODSEQ`. As of latest imap4.py (lastchangerev [18812]), this doesn't breaks IMAP4Client:

```
          File "/usr/lib64/python2.4/site-packages/twisted/mail/imap4.py", line 2300, in __cbCapabilities
            caps.setdefault(cap[:eq], []).append(cap[eq+1:])
        exceptions.AttributeError: 'NoneType' object has no attribute 'append'
```

Here's a fix, but I don't have unit tests right now. Note I decided to keep the cap unchanged if, say, `SORT=MODSEQ` precedes `SORT`.

```
--- imap4.py    (revision 20482)
+++ imap4.py    (working copy)
@@ -2295,7 +2295,7 @@
             for cap in rest:
                 eq = cap.find('=')
                 if eq == -1:
-                    caps[cap] = None
+                    caps[cap] = caps.get(cap, [])
                 else:
                     caps.setdefault(cap[:eq], []).append(cap[eq+1:])
         self._capCache = caps
```

Attachments:

* [imap4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/667/6676319c523479160a1952182ce1f8ce29b48b65/859a30cdb399d2559e889de25bb96ab68071ad92.diff) (803 bytes) - added by bkirsch on 2007-06-07 20:55:11Z - This is the same diff as the previous post except that is preserves the current external Twisted Capabilities API

<details><summary>Searchable metadata</summary>

```
trac-id__2695 2695
type__defect defect
reporter__slamb slamb
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1181185062000000 1181185062000000
changetime__1181576146000000 1181576146000000
version__None None
owner__exarkun exarkun
cc__slamb cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Fixed in imap4-mixed-capabilities-2695

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

I didn't spot any problem, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [20527]) Merge imap4-mixed-capabilities-2695

Author: exarkun
Reviewer: therve
Fixes [#2695](https://github.com/twisted/twisted/issues/2695)

Add support to IMAP4Client for handling CAPABILITY responses
which include atoms of the form "FOO" and "FOO=BAR" (in either
order).
